pkg/operator/controller/extension/virtualcluster/add.go

```diff
@@ -0,0 +1,111 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package virtualcluster
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/clock"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	operatorv1alpha1 "github.com/gardener/gardener/pkg/apis/operator/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	clientmapbuilder "github.com/gardener/gardener/pkg/client/kubernetes/clientmap/builder"
+	"github.com/gardener/gardener/pkg/controllerutils/mapper"
+	"github.com/gardener/gardener/pkg/operator/controller/garden/care"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "extension-virtual-cluster"
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(ctx context.Context, mgr manager.Manager) error {
+	if r.RuntimeClient == nil {
+		runtimeClientSet, err := kubernetes.NewWithConfig(
+			kubernetes.WithRESTConfig(mgr.GetConfig()),
+			kubernetes.WithRuntimeAPIReader(mgr.GetAPIReader()),
+			kubernetes.WithRuntimeClient(mgr.GetClient()),
+			kubernetes.WithRuntimeCache(mgr.GetCache()),
+		)
+		if err != nil {
+			return fmt.Errorf("failed creating runtime clientset: %w", err)
+		}
+		r.RuntimeClient = runtimeClientSet.Client()
+	}
+
+	if r.Clock == nil {
+		r.Clock = clock.RealClock{}
+	}
+	if r.Recorder == nil {
+		r.Recorder = mgr.GetEventRecorderFor(ControllerName + "-controller")
+	}
+	if r.GardenNamespace == "" {
+		r.GardenNamespace = v1beta1constants.GardenNamespace
+	}
+	if r.GardenClientMap == nil {
+		var err error
+		r.GardenClientMap, err = clientmapbuilder.
+			NewGardenClientMapBuilder().
+			WithRuntimeClient(r.RuntimeClient).
+			WithClientConnectionConfig(&r.Config.VirtualClientConnection).
+			WithGardenNamespace(r.GardenNamespace).
+			Build(mgr.GetLogger())
+		if err != nil {
+			return fmt.Errorf("failed to build garden ClientMap: %w", err)
+		}
+		if err := mgr.Add(r.GardenClientMap); err != nil {
+			return err
+		}
+	}
+
+	return builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(&operatorv1alpha1.Extension{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: ptr.Deref(r.Config.Controllers.ExtensionVirtualClusterConfig.ConcurrentSyncs, 0),
+		}).
+		Watches(
+			&operatorv1alpha1.Garden{},
+			mapper.EnqueueRequestsFrom(ctx, mgr.GetCache(), mapper.MapFunc(r.MapToAllExtensions), mapper.UpdateWithNew, mgr.GetLogger()),
+			builder.WithPredicates(predicate.Or(care.GardenPredicate(), predicate.GenerationChangedPredicate{})),
```

Comment:
I may be wrong but, couple of lines above we have the:
```
For(&operatorv1alpha1.Extension{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
```

so we should be "watching" on extension events. We watch the `Garden` objects in addition to the extension objects, because after all this controller does need a `Garden` ready before doing anything.

Reply:
> so we should be "watching" on extension events.

May bad, I missed that 👍 

> We watch the Garden objects in addition to the extension objects, because after all this controller does need a Garden ready before doing anything.

Hm, the reconciler already re-enqueues when the virtual Garden is not available. With that additional watch, two reconciliations in a row might happen which is probably not too critical and an acceptable trade-off.
However, why do we need to reconcile when the generation of the `Garden` changes?

Reply:
I removed the requeue error when the virtual cluster is not ready and only left the watch that re-enqueues when the garden's is ready. We need the second watch for the Garden anyway so we might as well rely on that.